### PR TITLE
Superglue part 1

### PR DESCRIPTION
**Pull request by iacolippo**
Added the stub of the tasks that are similar to GLUE (boolq, cb, rte, wic). The other tasks have different number of inputs, and different finetuning schemes. I added a comment about those in superglue/data.py

@NohTow I think I won't have time to go through all the details we need to get this right, if you could take it from here it would be awesome. You can maybe have help from other people in the team as well.

Notice that there is a lot of code that's duplicated right now, I didn't want to factorize it before everything was implemented so that we can make it make sense.

### CODE REVIEW

**Comment by jackcook** [comment]
I spoke with @NohTow and he invited me to work on this PR since he's been busy with other tasks. I restructured the evals files a bit to get rid of some redundant code, and then updated the `glue.py` script to accommodate the new SuperGLUE tasks that @iacolippo added. You can run it with:

```
python glue.py yamls/baselines/bert-base-uncased-superglue.yaml
```

And after about 2 hours on a 4xA40 machine it should output something along the lines of:

```
Collected Job Results: 

-------------------------------------------------------------
Task                                              |
-------------------------------------------------------------
MNLI                                              | 84.55
RTE                                               | 77.55
GLUE                                              | 81.05
-------------------------------------------------------------



Collected Job Results: 

-------------------------------------------------------------
Task                                              |
-------------------------------------------------------------
BOOLQ                                             | 73.41
CB                                                | 80.36
WIC                                               | 69.91
RTE                                               | 77.55
SUPERGLUE                                         | 75.31
-------------------------------------------------------------
```

There's some more work left to be done here, which I can tackle over the next few days, but am happy to pair on if anyone wants to help:
- [ ] Ensure these scores are good -- a couple of them are a few points off of the [BERT baseline](https://super.gluebenchmark.com/submission/G4txQjxKHkYv0QcBD06oq1KZPUw2/-LjNql9HhyWpGqIvNSVZ), but according to the SuperGLUE paper these are for BERT-large. SuperGLUE doesn't publish a BERT-base baseline, I'm sure a good one exists but I haven't looked for it
- [ ] Implement the remaining SuperGLUE tasks:
  - [ ] `axb`
  - [ ] `axg`
  - [ ] `copa`
  - [ ] `record`
  - [ ] `multirc`
  - [ ] `wsc`

Once all the tasks are in, the changes I made to `glue.py` should allow us to fit all the tasks in a single yaml to quickly and easily run and get scores for GLUE and SuperGLUE together, as shown above.

**Comment by jackcook** [comment]
Here are some results using the latest commit in this PR for `bert-base-uncased`: https://gist.github.com/jackcook/7c3a10f93460c0e074613b7d48e44fa0

This PR has gotten pretty big, so following some discussion today I'm going to open up this PR for review now since I think parts of it (especially the `glue` -> `evals` reorganization, the `tokenize_fn_factory` change, and the `collate_fn` change) will be useful in implementing other benchmarks. The remaining two SuperGLUE benchmarks (ReCoRD and WSC) can be implemented in a separate PR.

To make the review easier, here are some of the highlights:

- Implemented several benchmarks:
  - SuperGLUE metrics: BoolQ, CB, COPA, MultiRC, WiC (`src/evals/superglue_jobs.py`)
  - Other metrics: SWAG (used for COPA finetunes — `src/evals/misc_jobs.py`)
- Support reporting results from multiple benchmark suites (GLUE, SuperGLUE, others in the future) at the same time
  - `glue.py` [lines 539-580](https://github.com/AnswerDotAI/bert24/pull/34/files#diff-c159c90fa26b7e52212fa6a0b6f22664dd4b27aa08e6e65a8fd64fef2da01de1R539-R580)
- Renamed `glue` directory to `evals`
- Added multiple-choice variants of FlexBERT and MosaicBERT to support multiple choice benchmarks (SWAG, COPA)
  - `src/hf_bert.py` [lines 206-225](https://github.com/AnswerDotAI/bert24/pull/34/files#diff-488e3b01a57cd810a942a31512064cb41b607188bbe6d90e945988b4f58582d6R206-R225)
  - `src/mosaic_bert.py` [lines 255-275](https://github.com/AnswerDotAI/bert24/pull/34/files#diff-70a418e7feeb2f9539f70b39ff950988bfbdeff996317549b9b4a6ea1c3d4482R255-R275)
  - `src/flex_bert.py` [lines 259-280](https://github.com/AnswerDotAI/bert24/pull/34/files#diff-129a4174276fbbfe1fc0167934eefaac49fdb1d01113d89f1d9b049372d74213R259-R280)
- Enabled certain benchmarks to depend on models other than MNLI (COPA is much easier to finetune from a model finetuned on SWAG)
  - `glue.py` [lines 480-511](https://github.com/AnswerDotAI/bert24/pull/34/files#diff-c159c90fa26b7e52212fa6a0b6f22664dd4b27aa08e6e65a8fd64fef2da01de1R480-511)
- Enabled benchmarks to support custom metrics for those that don't simply rely on accuracy/F1 scores (CB, MultiRC)
  - `src/hf_bert.py` [line 255](https://github.com/AnswerDotAI/bert24/pull/34/files#diff-488e3b01a57cd810a942a31512064cb41b607188bbe6d90e945988b4f58582d6R255)
- Enabled advanced benchmarks to use their own tokenization and collation functions (SWAG, COPA, MultiRC)
  - `src/evals/data.py` [lines 98-107](https://github.com/AnswerDotAI/bert24/pull/34/files#diff-d07a0c8026f81ef5401c14c611311e50784378a309e79269985792c45dfddfe7R98-R107)
  - `src/evals/finetuning_jobs.py` [lines 67-69](https://github.com/AnswerDotAI/bert24/pull/34/files#diff-4e712b4f552bf50782daee44ac62ae57c1543620699bd4f4788aedff7340d8b2R67-R69)

**Comment by jackcook** [comment]
It's not letting me add you as a reviewer since you made the PR @iacolippo but I figure you might also want to take a look

File: src/bert_layers/model.py [hunk 1]
```diff
@@ -506,8 +565,120 @@ def forward(
...
+        self.config = config
+
+        self.bert = BertModel(config)
+        classifier_dropout = (
+            config.classifier_dropout
+            if config.classifier_dropout is not None
+            else config.hidden_dropout_prob
+        )
+        self.dropout = nn.Dropout(classifier_dropout)
+        self.classifier = nn.Linear(config.hidden_size, 1)
```
Comment (iacolippo) [inline src/bert_layers/model.py:585]: it took some time to me to figure out how this works (get logit for each candidate than normalize). The first thing I thought was that this should have been hidden_size --> num_labels 😓 a comment on how this type of head works would be good for posterity I think

File: src/bert_layers/model.py [hunk 2]
```diff
@@ -844,3 +1075,100 @@ def forward(
...
+        loss = None
+        if labels is not None:
+            loss_fct = nn.CrossEntropyLoss()
+            loss = loss_fct(reshaped_logits, labels)
+
+        if not return_dict:
+            output = (reshaped_logits,) + output
+            return ((loss,) + output) if loss is not None else output
+
+        return SequenceClassifierOutput(
```
Comment (iacolippo) [inline src/bert_layers/model.py:1169]: Why a different return type wrt BertForMultipleChoice? (sequence instead of multiplechoice out)

File: src/evals/data.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        download_config=download_config,
+    )
+
+    log.info(f"Starting tokenization by preprocessing over {num_workers} threads!")
+    text_column_names = task_column_names[task]
+
+    if tokenize_fn_factory is None:
+        tokenize_fn_factory = lambda tokenizer, max_seq_length: lambda inp: tokenizer(
+            text=inp[text_column_names[0]],
+            text_pair=(
```
Comment (iacolippo) [inline src/evals/data.py:101]: a comment explaning that this adds a [SEP] in between is welcome IMO 😄

Comment (iacolippo) [reply src/evals/data.py:101]: We should also mention that without NSP this token is not pretrained I guess but since we train first on MNLI and then use that checkpoint, we still get meaningful results.

**Review by iacolippo — COMMENTED** [review]
I haven't checked the details of each finetuning job yet but I like the factory approach for the tokenization, and I just added some comments where I think we should make clear why certain things are the way they are. I'll go over each finetuning job later.

Thank you for taking care of this btw!

File: src/bert_layers/model.py (hunk 2 above)
Comment (jackcook) [reply src/bert_layers/model.py:1169]: Good catch, my mistake

**Comment by rbiswasfc** [comment]
Really like structuring of the eval jobs -- makes addition of new evals a lot easier. I'm able to run `python glue.py yamls/baselines/bert-base-uncased-superglue.yaml` and reproduce the results [here](https://gist.github.com/jackcook/7c3a10f93460c0e074613b7d48e44fa0). I've two minor questions:
- how are the params (e.g. lr and max_duration) selected for different fine-tuning jobs? 
- Is there a specific reason for disabling shuffle in the training data loader? e.g.
```python
dataloader_kwargs = {
    "batch_size": self.batch_size,
    "num_workers": 0,
    "shuffle": False,
    "drop_last": False,
}
train_dataset = create_glue_dataset(split="train", **dataset_kwargs)
self.train_dataloader = build_dataloader(train_dataset, **dataloader_kwargs)
```

**Comment by bclavie** [comment]
Seconding the other comments, I really like the structure & it makes it really easy to extend. LGTM overall.

My only (nit:)question's the same as Raja: is there a specific reason we disable shuffling? I seem to remember that SuperGLUE's default is to shuffle but I might be wrong!

**Review by bclavie — APPROVED** [review]

**Comment by jackcook** [comment]
Good catch, I disabled it while doing some testing but we should turn it back on — let me fix that real quick along with this merge conflict

**Comment by jackcook** [comment]
And as for @rbiswasfc's comment, I tweaked them by hand until I got results that seemed roughly correct with what's expected for BERT. Doing some hyperparameter tuning to find better values would probably be a good task to complete at some point if we have enough time/compute.

**Comment by bclavie** [comment]
Also sorry forgot to respond to this @jackcook:

> Ensure these scores are good -- a couple of them are a few points off of the [BERT baseline](https://super.gluebenchmark.com/submission/G4txQjxKHkYv0QcBD06oq1KZPUw2/-LjNql9HhyWpGqIvNSVZ), but according to the SuperGLUE paper these are for BERT-large. SuperGLUE doesn't publish a BERT-base baseline, I'm sure a good one exists but I haven't looked for it

Surprisingly enough, I've never found a good report of full `bert-base` results on SuperGLUE. My go-to for sanity checks is generally from UKPLab's [work on selecting the optimal "middle-task" for fine-tuning](https://arxiv.org/abs/2104.08247) (BERT results are on the very last page). It does seem like BERT-base + MNLI should score _a bit_ higher than what we're seeing, but not sure what the issue could be.

**Comment by bclavie** [comment]
(although looking at the full results, it might well be that they had slightly luckier seeds than us, nothing seems _extremely _ out-of-the-ordinary)

**Comment by jackcook** [comment]
This PR definitely isn't training all of the tasks as long as they could be trained for -- there's almost certainly an opportunity to get another point or two here and there. Maybe we can merge this in for now and look at that for a future PR?

**Comment by bclavie** [comment]
Sounds good to me — was more of a nitpick than a requirement anyway! Thank you so much for all the work on this

**Comment by rbiswasfc** [comment]
> And as for @rbiswasfc's comment, I tweaked them by hand until I got results that seemed roughly correct with what's expected for BERT. Doing some hyperparameter tuning to find better values would probably be a good task to complete at some point if we have enough time/compute.

Noted, thanks! I think we should avoid tuning hyperparameters -- considering our goal is to access transfer learning potential of the base models. Tuning HPs will most likely lead to over optimistic eval scores, magnify the influence of fine-tuning + compute cost of a sweep can be prohibitive...

**Comment by jackcook** [comment]
Ah to be clear, I didn't tune them further once they got within a couple points of the SuperGLUE baselines. I just tuned them to the point where they would learn reasonably well, and then stopped -- for some of the tasks, they wouldn't get learned at all with certain parameters.

**Comment by rbiswasfc** [comment]
> Ah to be clear, I didn't tune them further once they got within a couple points of the SuperGLUE baselines. I just tuned them to the point where they would learn reasonably well, and then stopped -- for some of the tasks, they wouldn't get learned at all with certain parameters.

Perfect, I also think this approach works best in this case!
